### DISCUSSION TITLE
Preferred path for making simple UI's

### DISCUSSION THREAD

**Discussion by burbma** [gokrazy/gokrazy#249] [Q&A] [answered]
Allow me to tell a short story so that with your expertise you can answer the question I should've asked but am too novice to know to ask. I've been looking at making a simple computing device for my kids, I started off by running Raspberry OS Lite on my Pi and wiring it up to boot straight into [fbterm](https://github.com/sfzhi/fbterm) running a Python [Textual](https://github.com/Textualize/textual) app. The fbterm layer allowed me to use nice fonts. This was still more bloated and round about than I wanted but seemed preferable over making a chrome kiosk. It also left me with no updating solutions out of the box. When seeking to make my program faster I considered using Go and I found [Bubble Tea](https://github.com/charmbracelet/bubbletea/tree/master) which would replace Textual, though I've never used Go until this point. Then, since I was in Go land at that point I discovered gokrazy which seems awesome as it has some solutions to both the bloat and updating. Would using Bubble Tea inside a gokrazy appliance make sense? I gave it a naive try with the result being "Can't find /dev/tty" in the logs shown in the gokrazy UI. Or is the approach used for fbstatus the best way to create a UI?

**Comment by stapelberg** [comment, marked as answer]
Hey @burbma 👋

Bubble Tea is a package that generates textual user interfaces via terminal escape sequences.

If I understand correctly, you’re trying to have your Raspberry Pi boot into a textual interface where a Bubble Tea app is running.

To accomplish this, you’re trying to use the Linux Virtual Console (https://en.wikipedia.org/wiki/Virtual_console), which seems like a good choice to me.

Bubble Tea complains about /dev/tty missing because gokrazy does not start its programs with a controlling tty (see https://unix.stackexchange.com/questions/229530/how-linux-uses-dev-tty-and-dev-tty0), which is intentional.

I think the component you need here is a replacement for https://en.wikipedia.org/wiki/Getty_(Unix), which should create a controlling tty when you interactively use the Linux Virtual Console. Note: Once you have a controlling tty, I’m not entirely sure how much of Bubble Tea the Linux Virtual Console supports. It would be easiest to try out on a non-gokrazy system beforehand.

I don’t use the Linux Virtual Console with gokrazy typically (I use SSH or the serial port), but it should be possible to make it work. I hope this gives you a couple of pointers to pursue

**Reply by burbma** [reply]
This is excellent @stapelberg thank you! This gives me a better vocabulary and starting resources to learn more. I'll dig in!

**Reply by burbma** [reply]
@stapelberg I was able to get an example Bubbletea app to display and function on the Linux Virtual Console by copying some of the code I found in [fbstatus/internal/console/console.go](https://github.com/gokrazy/fbstatus/blob/main/internal/console/console.go) and using Bubbletea's [WithInput](https://pkg.go.dev/github.com/charmbracelet/bubbletea#WithInput) and [WithOutput](https://pkg.go.dev/github.com/charmbracelet/bubbletea#WithOutput) as seen in [this gist](https://gist.github.com/burbma/7ca9c91c5f59afa062d293235fa2126b#file-vt_tea-go-L52). But as you suggested it doesn't look very nice or function perfectly smoothly in the Linux Virtual Console. Do you have any tips on how I might get this improved? Maybe with some kind of framebuffer layer or running a more modern terminal emulator somehow?

**Reply by stapelberg** [reply]
Can you provide functioning code please so that I can try it out and better understand what exactly isn’t running smoothly?

**Reply by betim** [reply]
Is there a way to boot a chrome in kiosk mode with gokrazy?

**Reply by damdo** [reply]
See: https://github.com/damdo/gokrazy-kiosk

**Reply by betim** [reply]
> See: https://github.com/damdo/gokrazy-kiosk

Thank you my friend. I tried googling a bit before asking the question but did not find anything. Maybe you should mention chrome in README.md or maybe I should have searched chromium instead of chrome, but you know, one sometimes forgets.

Thanks for you effort!
